pep-0590.rst

```diff
@@ -48,145 +48,212 @@ The function pointer type
 
 New C API and changes to CPython
 ================================
 
-``PyObject *PyObject_VectorCall(PyObject *obj, PyObject **args, Py_ssize_t nargs, PyTupleObject *kwnames)``
+The following functions or macros are added to the C API:
 
-Calls ``obj`` with the given arguments. Note that ``nargs`` may include the flag ``PY_VECTORCALL_ARGUMENTS_OFFSET``.
-``nargs & ~PY_VECTORCALL_ARGUMENTS_OFFSET`` is the number of positional arguments.
+- ``PyObject *PyObject_Vectorcall(PyObject *obj, PyObject **args, Py_ssize_t nargs, PyObject *kwargs)``:
+  Calls ``obj`` with the given arguments.
+  Note that ``nargs`` may include the flag ``PY_VECTORCALL_PREPEND``.
+  The actual number of positional arguments is given by ``PyVectorcall_NARGS(nargs)``.
+  The argument ``kwargs`` is either a dict, a tuple of keyword names or ``NULL``.
+  An empty dict or empty tuple has the same effect as passing ``NULL``.
+  This uses either the vectorcall protocol or ``tp_call`` internally;
+  if neither is supported, an exception is raised.
 
-``PyObject_VectorCall`` raises an exception if ``obj`` is not callable.
+- ``int PyVectorcall_Check(PyObject *obj)``: return 1 if ``obj`` supports the vectorcall
+  protocol, 0 otherwise.
 
-Two utility functions are provided to call the new calling convention from the old one, or vice-versa.
-These functions are ``PyObject *PyCall_MakeVectorCall(PyObject *obj, PyObject *tuple, PyObject **dict)`` and
-``PyObject *PyCall_MakeTpCall(PyObject *obj, PyObject **args, Py_ssize_t nargs, PyTupleObject *kwnames)``, respectively.
+- ``PyObject *PyVectorcall_Call(PyObject *obj, PyObject *tuple, PyObject **dict)``:
+  Call the object (which must support the vectorcall protocol) with the old
+  ``*args`` and ``**kwargs`` calling convention.
+  This is mostly meant to put in the ``tp_call`` slot.
 
-Both functions raise an exception if ``obj`` does not support the relevant protocol.
+- ``Py_ssize_t PyVectorcall_NARGS(Py_ssize nargs)``: currently equivalent to
+  ``nargs & ~PY_VECTORCALL_PREPEND``.
 
-``METH_FASTCALL`` and ``METH_VECTORCALL`` flags
------------------------------------------------
+New ``METH_VECTORCALL`` flag
+----------------------------
 
-A new ``METH_VECTORCALL`` flag is added for specifying ``PyMethodDef`` structs. It is equivalent to the currently undocumented ``METH_FASTCALL | METH_KEYWORD`` flags.
-The new flag specifies that the function has the type ``PyObject *(*call) (PyObject *self, PyObject *const *args, Py_ssize_t nargs, PyObject *kwname)``
+A new constant ``METH_VECTORCALL`` is added for specifying ``PyMethodDef`` structs.
+It means that the C function has the signature ``vectorcallfunc``.
+This should be the preferred flag for new functions, as this avoids a wrapper function.
```

Comment:
This is a major change and prevents the interpreter from sanity checking the return value of `PyMethodDef` defined functions. The signature of `METH_VECTORCALL` should remain as
`PyObject *(*call) (PyObject *self, PyObject *const *args, Py_ssize_t nargs, PyObject *kwname)`

Reply:
> This is a major change and prevents the interpreter from sanity checking the return value of `PyMethodDef` defined functions.

Please explain. I don't think it's the job of the vectorcall **callee** to do that check, it should be done by the vectorcall **caller**.

Reply:
No, it is the job of the callee to do this check. `_PyCFunction_FastCallKeywords` does this, as the `PymethodDef` defined code is not entirely trusted. However, calls to the interpreter or builtin classes *are* trusted and don't need to be double checked.

Reply:
> as the `PymethodDef` defined code is not entirely trusted.

Why should the `PyMethodDef`-defined code be less trusted than other code called by the vectorcall protocol? I don't see the difference.

Reply:
Because `PyMethodDef` remains the preferred way for builtin functions to be implemented if they are implemented in C. For tools like Cython that can ensure that the returned value is consistent, the additional check is not required.